File: DSA/Manav-Rajput/Day51_question2.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int left = getMax(nums); // minimum possible largest sum
        int right = getSum(nums); // maximum possible largest sum
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (canSplit(nums, k, mid)) {
                right = mid; // mid might be the answer, so move right boundary to mid
            } else {
                left = mid + 1; // mid is too small, increase left boundary
            }
        }
        
        return left;
    }
    
    // Helper function to find the maximum element in the array
    private int getMax(int[] nums) {
        int max = nums[0];
        for (int num : nums) {
            max = Math.max(max, num);
        }
        return max;
    }
    
    // Helper function to find the sum of elements in the array
    private int getSum(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        return sum;
    }
    
    // Helper function to check if we can split the array into `k` or fewer subarrays
    // with each subarray having a sum less than or equal to `maxSum`
    private boolean canSplit(int[] nums, int k, int maxSum) {
        int currentSum = 0;
        int splits = 1;
        
        for (int num : nums) {
            if (currentSum + num > maxSum) {
                splits++;
                currentSum = num;
                if (splits > k) {
                    return false;
                }
            } else {
                currentSum += num;
            }
        }
        
        return true;
    }
}

File: DSA/Arshdeep-Kaur/Day 51/Day51_Question2.cpp
class Solution {
public:
    int countSubarrays(vector<int> &nums, int k) {
        int subarrays = 1;
        long long ans = 0;
        for(int i=0; i<nums.size(); i++) {
            if (ans + nums[i] <= k) {
                ans += nums[i];
            }
            else {
                subarrays += 1;
                ans = nums[i];
            }
        }
        return subarrays;
    }

    int splitArray(vector<int>& nums, int k) {
        int low = *max_element(nums.begin(), nums.end());
        int high = accumulate(nums.begin(), nums.end(), 0);

        while(low <= high) {
            int mid = (low + high) / 2;
            int subarrays = countSubarrays(nums, mid);
            if (subarrays > k) {
                low = mid + 1;
            }
            else {
                high = mid - 1;
            }
        }

        return low;
    }
};
